File: ADP.Surveys/ADP.Surveys.Shared/Evaluation/LogicEvaluator.cs
using System.Text.Json;
using ShiftSoftware.ADP.Surveys.Shared.DTOs;
using ShiftSoftware.ADP.Surveys.Shared.DTOs.Logic;
using ShiftSoftware.ADP.Surveys.Shared.Enums;
using ShiftSoftware.ADP.Surveys.Shared.Evaluation.ExpressionSandbox;

namespace ShiftSoftware.ADP.Surveys.Shared.Evaluation;

/// <summary>
/// Evaluates cross-screen logic rules against an accumulated answer map. Per the
/// schema contract, rules are ordered; the first one whose condition matches wins.
/// If none match, the caller falls back to <c>screen.NextScreen</c> (or sequential
/// order, or end of survey).
///
/// Per Decision #10: evaluator accepts both structured conditions (predicates +
/// <c>all</c>/<c>any</c> composites) and the string-expression escape hatch
/// (<see cref="ExpressionConditionDto"/>), delegating the latter to <see cref="ExpressionSandbox.ExpressionSandbox"/>.
/// Any runtime evaluation failure is treated as <c>false</c> so a broken rule
/// falls through to the default navigation rather than blocking the survey.
/// </summary>
public static class LogicEvaluator
{
    /// <summary>
    /// Walks <see cref="SurveyDto.Logic"/> in order; returns the first matching rule's
    /// goto target, or <c>null</c> when nothing matches.
    /// </summary>
    public static string? EvaluateNext(SurveyDto survey, IReadOnlyDictionary<string, JsonElement> answers)
    {
        foreach (var rule in survey.Logic)
        {
            if (EvaluateCondition(rule.If, answers))
                return rule.Then?.Goto;
        }
        return null;
    }

    /// <summary>
    /// Evaluate a single condition. Exposed so the builder's rule-preview panel
    /// can test rules one at a time.
    /// </summary>
    public static bool EvaluateCondition(LogicConditionDto condition, IReadOnlyDictionary<string, JsonElement> answers)
    {
        try
        {
            return condition switch
            {
                PredicateConditionDto p => EvaluatePredicate(p, answers),
                CompositeConditionDto c => EvaluateComposite(c, answers),
                ExpressionConditionDto e => ExpressionSandbox.ExpressionSandbox.EvaluateBoolean(e.Expression, answers),
                _ => false,
            };
        }
        catch
        {
            // Defensive: a broken rule must not block navigation. Caller falls through.
            return false;
        }
    }

    private static bool EvaluateComposite(CompositeConditionDto composite, IReadOnlyDictionary<string, JsonElement> answers)
    {
        if (composite.All is { Count: > 0 } all)
            return all.All(child => EvaluateCondition(child, answers));
        if (composite.Any is { Count: > 0 } any)
            return any.Any(child => EvaluateCondition(child, answers));
        return false;
    }

    private static bool EvaluatePredicate(PredicateConditionDto predicate, IReadOnlyDictionary<string, JsonElement> answers)
    {
        var answerPresent = answers.TryGetValue(predicate.QuestionId, out var answerRaw)
            && answerRaw.ValueKind != JsonValueKind.Null;

        switch (predicate.Op)
        {
            case LogicOperator.IsSet: return answerPresent;
            case LogicOperator.IsNotSet: return !answerPresent;
        }

        // The remaining operators all need a value; treat "no value provided" as non-matching.
        if (!predicate.Value.HasValue) return false;

        var answerValue = answerPresent ? JsonValueHelper.Unwrap(answerRaw) : null;
        var predicateValue = JsonValueHelper.Unwrap(predicate.Value.Value);

        return predicate.Op switch
        {
            LogicOperator.Equals => JsonValueHelper.AreEqual(answerValue, predicateValue),
            LogicOperator.NotEquals => !JsonValueHelper.AreEqual(answerValue, predicateValue),
            LogicOperator.GreaterThan => JsonValueHelper.Compare(answerValue, predicateValue) > 0,
            LogicOperator.GreaterThanOrEqual => JsonValueHelper.Compare(answerValue, predicateValue) >= 0,
            LogicOperator.LessThan => JsonValueHelper.Compare(answerValue, predicateValue) < 0,
            LogicOperator.LessThanOrEqual => JsonValueHelper.Compare(answerValue, predicateValue) <= 0,
            LogicOperator.In => InList(predicateValue, answerValue),
            LogicOperator.NotIn => !InList(predicateValue, answerValue),
            _ => false,
        };
    }

    private static bool InList(object? list, object? value)
    {
        if (list is not List<object?> items) return false;
        return items.Any(item => JsonValueHelper.AreEqual(value, item));
    }
}

File: ADP.Surveys/renderer/packages/survey-sdk/src/logic-evaluator.ts
/**
 * Cross-screen logic evaluator. Mirror of
 * `ADP.Surveys.Shared/Evaluation/LogicEvaluator.cs`.
 *
 * Walks `survey.logic` in order and returns the first matching rule's goto target,
 * or null when nothing matches. Caller falls back to `screen.nextScreen` / sequential
 * / end-of-survey per the Phase 3 navigation engine.
 *
 * Decision #10 discipline: broken rules (unresolvable predicate types, compare-on-
 * mismatched-types, malformed expressions) fall through as false so survey
 * navigation always progresses.
 */

import type {
  AnswerMap,
  CompositeCondition,
  LogicCondition,
  LogicOperator,
  PredicateCondition,
  Survey,
} from './schema.js';
import { isCompositeCondition, isExpressionCondition, isPredicateCondition } from './schema.js';
import { evaluateBoolean as evaluateExpression } from './expression-sandbox/index.js';
import { areEqual, compare, unwrap } from './expression-sandbox/value-helper.js';

export function evaluateNext(survey: Survey, answers: AnswerMap): string | null {
  if (!survey.logic) return null;
  for (const rule of survey.logic) {
    if (evaluateCondition(rule.if, answers)) return rule.then?.goto ?? null;
  }
  return null;
}

export function evaluateCondition(condition: LogicCondition, answers: AnswerMap): boolean {
  try {
    if (isPredicateCondition(condition)) return evaluatePredicate(condition, answers);
    if (isCompositeCondition(condition)) return evaluateComposite(condition, answers);
    if (isExpressionCondition(condition)) return evaluateExpression(condition.expression, answers);
    return false;
  } catch {
    // Defensive: a broken rule must not block navigation. Caller falls through.
    return false;
  }
}

function evaluateComposite(composite: CompositeCondition, answers: AnswerMap): boolean {
  if (composite.all && composite.all.length > 0) {
    return composite.all.every((child) => evaluateCondition(child, answers));
  }
  if (composite.any && composite.any.length > 0) {
    return composite.any.some((child) => evaluateCondition(child, answers));
  }
  return false;
}

function evaluatePredicate(predicate: PredicateCondition, answers: AnswerMap): boolean {
  const answerPresent =
    predicate.questionId in answers &&
    answers[predicate.questionId] !== null &&
    answers[predicate.questionId] !== undefined;

  if (predicate.op === 'isSet') return answerPresent;
  if (predicate.op === 'isNotSet') return !answerPresent;

  // Remaining operators need a value; treat "no value provided" as non-matching.
  if (predicate.value === undefined) return false;

  const answerValue = answerPresent ? unwrap(answers[predicate.questionId]) : null;
  const predicateValue = unwrap(predicate.value);

  return applyOperator(predicate.op, answerValue, predicateValue);
}

function applyOperator(op: LogicOperator, answer: unknown, value: unknown): boolean {
  switch (op) {
    case '==':
      return areEqual(answer, value);
    case '!=':
      return !areEqual(answer, value);
    case '>':
      return compare(answer, value) > 0;
    case '>=':
      return compare(answer, value) >= 0;
    case '<':
      return compare(answer, value) < 0;
    case '<=':
      return compare(answer, value) <= 0;
    case 'in':
      return inList(value, answer);
    case 'notIn':
      return !inList(value, answer);
    default:
      return false;
  }
}

function inList(list: unknown, value: unknown): boolean {
  if (!Array.isArray(list)) return false;
  return list.some((item) => areEqual(value, item));
}
